### ISSUE TITLE
Reading more than 0x7fffffff bytes at once on macos fails

### ISSUE THREAD

**Issue by gaultier** [odin-lang/Odin#2098]
## Context

I hit this issue while trying to read a big file (0x911FF000 bytes), which is bigger than  0x7fffffff which is a hidden limit in the `read(2)` system call on macos.

Similar to https://github.com/ziglang/zig/pull/6333. Other languages have also hit this weird issue.

This triggers when using `os.read_at_least` and the functions calling it e.g. `os.read_full`, etc.

The link above has a fix which could be also done here meaning reading at most 0x7fffffff bytes at once on macos.

* Operating System & Odin Version:

```
        Odin: dev-2022-09:74458ab0
        OS:   macOS 12.0.6 Monterey
        CPU:  Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
        RAM:  16384 MiB
```

## Expected Behavior

`os.read_full` works on big files (bigger than 0x7fffffff).

## Current Behavior

We get an error (EINVAL) coming from the OS on macos.

### Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. Create a big enough file: ` head -c 2147483648 < /dev/zero > /tmp/test.txt`
2. Create a odin program to read the file:
```odin
package main

import "core:os"
import "core:fmt"

main :: proc() {
    if _, ok := os.read_entire_file_from_filename("/tmp/test.txt"); !ok {
      fmt.eprintf("Error: %s", os.get_last_error_string())
    }
}
```

4. Output: `Error: Invalid argument`
5. Confirming with dtruss: we get `read(0x3, "\0", 0x80000000)               = -1 Err#22` which is EINVAL

**Comment by gaultier** [comment]
This really is a kernel issue, the same can be seen only using C:

```c
#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <unistd.h>

int main() {
  uint64_t size = 1 << 31;
  char* buf = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_ANON, 0, 0);
  assert(buf != NULL);

  int fd = open("/tmp/big.txt", O_RDONLY, 0666);
  assert(fd >= 0);

  int ret = read(fd, buf, size);

  printf("%d %d\n", ret, errno);
}

```

**Comment by laytan** [comment]
Fixed in #2998
